### PR TITLE
Topic/index seq query

### PR DESCRIPTION
**Pull request by bryce-anderson**
This isn't quite finished, but I'd like to solicit opinions, suggestions, etc.
### Design
* Eagerly parsed. Initial benchmarks suggest its not a big penalty.
* ```Query``` is a wrapper around a ```Vector[(String, Option[String])]``` and extends the stdlib ```IndexedSeq``` trait. We could make a custom rolled structure ala spray. I'm still considering this.
* The api used on the ```Uri``` is shared with the ```Query``` through the ```QueryOps``` trait.

Generally, I like it, but I can forsee some performance impact on heavy query modification. Currently, the methods in ```QueryOps``` build an intermediate ```Map[String, Seq[String]]```, modify it, and then move back to a ```Query```. Adding multiple params in a row will be pretty nasty, eg
```scala
query +? (1k, 2v) +? (k2, v2) ...
```
I might spend a little time avoiding the intermediate map structure.

### CODE REVIEW

**Comment by rossabaker** [comment]
I think `Query.paramsView` should be just `Query.params` now.  But I like it overall.

**Comment by bryce-anderson** [comment]
I renamed ```paramsView``` to ```params``` and ```fromOptions``` to ```apply```, consistent with other collections constructors. 
I'd like to merge this, so I'd appreciate a thumbs up or suggestions for improvement.

File: core/src/test/scala/org/http4s/QuerySpec.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  checkAll("Query parsing", new Properties("Query") {
+    property("fromString(query.toString) == query if query.nonEmpty") = forAll { query: Query =>
+      (query.nonEmpty) ==> (Query.fromString(query.toString) == query)
+    }
+  })
+
+  "Query" can {
+    val elem = ("k", Some("v"))
+
+    "append a single Pair element" in {
```
Comment (julien-truffaut) [inline core/src/test/scala/org/http4s/QuerySpec.scala]: you could write these tests as scalacheck properties if you have time

Comment (bryce-anderson) [inline core/src/test/scala/org/http4s/QuerySpec.scala]: Done. Its kind of amazing how nice scalacheck works with specs2.

**Comment by bryce-anderson** [comment]
Alright, any other comments on this before I merge it? Its a "speak now, or we can make changes in another PR" situation.

**Comment by rossabaker** [comment]
The niceness of that integration was motivating factor in the specs2
migration.  See also the Scalaz support. :-)

**Comment by bryce-anderson** [comment]
I wonder what will become of that spec2 scalaz support in the coming months.

**Pull request merged**